### DISCUSSION TITLE
JsonType is undefined, global object in block content annotations

### DISCUSSION THREAD

**Discussion by JanneHelewaut** [sanity-io/sanity#3686] [Sanity Studio Support] [answered]
Hello!
I’m editing the annotations in the block content to support external links and other custom objects in the new V3.
Since I’m using GraphQL, I can’t have any inline objects and all objects need to be defined globally. So I created a type externalLink as:
```
export default {
  name: 'externalLink',
  title: 'External Link',
  type: 'object',
  liveEdit: false, 
  fields: [
    {
      title: 'URL',
      name: 'href',
      type: 'url',
    },
  ]
}
```
I added this type to the schema. but when I want to place it in the annotations of my block Content, I get the following error:
```
Error: Could not resolve jsonType of undefined. No parent type found
Error: Could not resolve jsonType of undefined. No parent type found
    at isJSONTypeOf (http://localhost:3333/node_modules/.vite/deps/chunk-TA2UENNL.js?v=47494152:83241:11)
    at http://localhost:3333/node_modules/.vite/deps/chunk-TA2UENNL.js?v=47494152:83366:24
    at Array.map (<anonymous>)
    at validateAnnotations (http://localhost:3333/node_modules/.vite/deps/chunk-TA2UENNL.js?v=47494152:83357:22)
    at validateMarks (http://localhost:3333/node_modules/.vite/deps/chunk-TA2UENNL.js?v=47494152:83306:19)
    at validateBlockType (http://localhost:3333/node_modules/.vite/deps/chunk-TA2UENNL.js?v=47494152:83271:13)
    at visitors.reduce._problems (http://localhost:3333/node_modules/.vite/deps/chunk-TA2UENNL.js?v=47494152:83585:19)
    at Array.reduce (<anonymous>)
    at http://localhost:3333/node_modules/.vite/deps/chunk-TA2UENNL.js?v=47494152:83584:21
    at http://localhost:3333/node_modules/.vite/deps/chunk-TA2UENNL.js?v=47494152:83599:40
```
And this is my BlockEditor code:
```
export default {
  title: 'Block Editor',
  name: 'blockEditor',
  type: 'array',
  of: [
    {
      title: 'Block',
      type: 'block',
      styles: [
        {title: 'Normal', value: 'normal'},
        {title: 'H2', value: 'h2'},
        {title: 'H3', value: 'h3'},
        {title: 'Quote', value: 'blockquote'},
      ],
      lists: [{title: 'Bullet', value: 'bullet'}],
      marks: {
        decorators: [
          {title: 'Strong', value: 'strong'},
          {title: 'Emphasis', value: 'em'},
        ],
        annotations: [
          {
            type: 'externalLink'
          }
        ],
      },
    },
    {
      type: 'image',
      name: 'image',
      options: {
        hotspot: true,
      }
    },
    {
      name: 'line',
      title: 'Line',
      type: 'line',
    },
  ],
}
```
[This example](https://github.com/sanity-io/sanity-template-nextjs-landing-pages/blob/main/template/studio/schemas/objects/portableText.js) shows me that it is possible (in V2), so my question is, what am I doing wrong here?
[I’m using version 3.0.1-dev-preview.0]

**Comment by skogsmaskin** [comment]
@JanneHelewaut - have you defined the externalLink type in the schema before it is used here?

The shorthand `{type: 'externalLink'}` would require that type to be defined elsewhere before it is used like that here.

Just to confirm this first.

**Reply by JanneHelewaut** [reply]
I did, if I use the externalLink outside of annotations it works fine. e.g. if I put it underneath the image and line, it is shown in the block content without any issues.

**Comment by JanneHelewaut** [comment]
I created a git repo with a really basic blockeditor that shows the issue really well.
https://github.com/JanneHelewaut/blockeditor

**Comment by JanneHelewaut** [comment, marked as answer]
After a lot of testing, we figured out the issue was in the order of the schemaTypes. When we put the externalLink before the BlockContent in the schemaTypes, it works.

**Comment by shaunchander** [comment]
We just ran into this issue earlier today, surely this isn't intentional behavior? The order of the schema types should not affect inter-referencing between schema types?
